BMS/STM32F767ZITX/Core/Src/freertos.c

```diff
@@ -0,0 +1,245 @@
+  .name = "TempSem"
+};
+
+/* Private function prototypes -----------------------------------------------*/
+/* USER CODE BEGIN FunctionPrototypes */
+
+/* USER CODE END FunctionPrototypes */
+
+void StartDefaultTask(void *argument);
+void StartTemp_Sense(void *argument);
+void StartCan_Transmitter(void *argument);
+
+void MX_FREERTOS_Init(void); /* (MISRA C 2004 rule 8.1) */
+
+/**
+  * @brief  FreeRTOS initialization
+  * @param  None
+  * @retval None
+  */
+void MX_FREERTOS_Init(void) {
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* USER CODE BEGIN RTOS_MUTEX */
+  /* add mutexes, ... */
+  /* USER CODE END RTOS_MUTEX */
+
+  /* Create the semaphores(s) */
+  /* creation of TempSem */
+  TempSemHandle = osSemaphoreNew(1, 0, &TempSem_attributes);
+
+  /* USER CODE BEGIN RTOS_SEMAPHORES */
+  /* add semaphores, ... */
+  /* USER CODE END RTOS_SEMAPHORES */
+
+  /* USER CODE BEGIN RTOS_TIMERS */
+  /* start timers, add new ones, ... */
+  /* USER CODE END RTOS_TIMERS */
+
+  /* Create the queue(s) */
+  /* creation of temp_sensor_data */
+  temp_sensor_dataHandle = osMessageQueueNew (16, sizeof(uint32_t), &temp_sensor_data_attributes);
+
+  /* USER CODE BEGIN RTOS_QUEUES */
+  /* add queues, ... */
+  /* USER CODE END RTOS_QUEUES */
+
+  /* Create the thread(s) */
+  /* creation of defaultTask */
+  defaultTaskHandle = osThreadNew(StartDefaultTask, NULL, &defaultTask_attributes);
+
+  /* creation of temp_sense */
+  temp_senseHandle = osThreadNew(StartTemp_Sense, NULL, &temp_sense_attributes);
+
+  /* creation of can_transmitter */
+  can_transmitterHandle = osThreadNew(StartCan_Transmitter, NULL, &can_transmitter_attributes);
+
+  /* USER CODE BEGIN RTOS_THREADS */
+  /* add threads, ... */
+  /* USER CODE END RTOS_THREADS */
+
+  /* USER CODE BEGIN RTOS_EVENTS */
+  /* add events, ... */
+  /* USER CODE END RTOS_EVENTS */
+
+}
+
+/* USER CODE BEGIN Header_StartDefaultTask */
+/**
+  * @brief  Function implementing the defaultTask thread.
+  * @param  argument: Not used
+  * @retval None
+  */
+/* USER CODE END Header_StartDefaultTask */
+void StartDefaultTask(void *argument)
+{
+  /* USER CODE BEGIN StartDefaultTask */
+  /* Infinite loop */
+  for(;;)
+  {
+    osDelay(1);
+  }
+  /* USER CODE END StartDefaultTask */
+}
+
+/* USER CODE BEGIN Header_StartTemp_Sense */
+/**
+* @brief Function implementing the temp_sense thread.
+* @param argument: Not used
+* @retval None
+*/
+/* USER CODE END Header_StartTemp_Sense */
+void StartTemp_Sense(void *argument)
+{
+  /* USER CODE BEGIN StartTemp_Sense */
+	uint32_t adc_data[NUM_MUX];
+	uint32_t temp_average[NUM_MUX] = {0}; //average temp from the previous iterations
+	uint32_t temp_data[NUM_SAMPLES][NUM_MUX] = {0};
+  /* Infinite loop */
+  for(;;)
+  {
+	//Measures average thermistor temperature of each MUX
+	measureTempADC(temp_average, adc_data, temp_data);
+	osMessageQueuePut(temp_sensor_dataHandle, temp_average, 0, osWaitForever);
```

Comment:
//Checks if mailbox is available for transmission and sends frame
if (HAL_CAN_GetTxMailboxesFreeLevel(&hcan3)) {
			CAN_send_frame(start_frame);
}